staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook_test.go

```diff
@@ -0,0 +1,353 @@
+	s := json.NewSerializer(json.DefaultMetaFactory, audit.Scheme, audit.Scheme, false)
+	return &testWebhookHandler{
+		t:          t,
+		onEvents:   cb,
+		serializer: s,
+	}
+}
+
+type testWebhookHandler struct {
+	t *testing.T
+
+	onEvents func(events *auditv1alpha1.EventList)
+
+	serializer runtime.Serializer
+}
+
+func (t *testWebhookHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	err := func() error {
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			return fmt.Errorf("read webhook request body: %v", err)
+		}
+
+		obj, _, err := t.serializer.Decode(body, nil, &auditv1alpha1.EventList{})
+		if err != nil {
+			return fmt.Errorf("decode request body: %v", err)
+		}
+		list, ok := obj.(*auditv1alpha1.EventList)
+		if !ok {
+			return fmt.Errorf("expected *v1alpha1.EventList got %T", obj)
+		}
+		t.onEvents(list)
+		return nil
+	}()
+
+	if err == nil {
+		io.WriteString(w, "{}")
+		return
+	}
+	// Must call Error instead of require.NoError or Fatal, because we're in
+	// a goroutine.
+	t.t.Error(err)
+	http.Error(w, err.Error(), http.StatusInternalServerError)
+}
+
+func newWebhook(t *testing.T, endpoint string, mode string) *AuditBackend {
+	config := v1.Config{
+		Clusters: []v1.NamedCluster{
+			{Cluster: v1.Cluster{Server: endpoint, InsecureSkipTLSVerify: true}},
+		},
+	}
+	f, err := ioutil.TempFile("", "k8s_audit_webhook_test_")
+	require.NoError(t, err, "creating temp file")
+
+	defer func() {
+		f.Close()
+		os.Remove(f.Name())
+	}()
+
+	// NOTE(ericchiang): Do we need to use a proper serializer?
+	require.NoError(t, stdjson.NewEncoder(f).Encode(config), "writing kubeconfig")
+
+	backend, err := NewBackend(f.Name(), mode)
+	require.NoError(t, err, "initializing backend")
+
+	return backend
+}
+
+func TestWebhook(t *testing.T) {
+	gotEvents := false
+	defer func() { require.True(t, gotEvents, "no events received") }()
+
+	s := httptest.NewServer(newWebhookHandler(t, func(events *auditv1alpha1.EventList) {
+		gotEvents = true
+	}))
+	defer s.Close()
+
+	backend := newWebhook(t, s.URL, ModeBlocking)
+
+	// Ensure this doesn't return a serialization error.
+	event := &auditinternal.Event{}
+	require.NoError(t, backend.sendEvents(event), "failed to send events")
+}
+
+func TestBatchWebhookMaxEvents(t *testing.T) {
+	events := make([]*auditinternal.Event, defaultBatchMaxSize+1) // greater than max size.
+	for i := range events {
+		events[i] = &auditinternal.Event{}
+	}
+
+	nRecieved := 0
+	defer func() {
+		require.Equal(t, defaultBatchMaxSize, nRecieved, "expected number of events did not equal number recieved")
+	}()
+
+	s := httptest.NewServer(newWebhookHandler(t, func(events *auditv1alpha1.EventList) {
+		nRecieved += len(events.Items)
+	}))
+	defer s.Close()
+
+	backend := newWebhook(t, s.URL, ModeBatch)
+
+	backend.queueEvents(events...)
+
+	stopCh := make(chan struct{})
+	timer := make(chan time.Time)
+
+	require.NoError(t, <-backend.sendBatchEvents(stopCh, timer), "failed to send batch events")
```

Comment:
Trying to write this test, there's actually a race here between reading from the buffer and observing the timer timeout. The timer send might be read before the events buffer reads that last event.